### DISCUSSION TITLE
delay attribute not working in commons-scxml2-2.0-SNAPSHOT version

### DISCUSSION THREAD

**Discussion by seenivasanmca** [alexzhornyak/SCXML-tutorial#14] [Q&A]
Unable to evaluate the expression from the .json from src tag.It is treating like string
{
"dp1":'10 &lt; 25',
} 

Delay attribute is not working properly. Here i have configured 5s delay but its is executing only once.

 

<scxml datamodel="jexl" location="Scxml" version="1.0"
    xmlns="http://www.w3.org/2005/07/scxml">

 

    <datamodel>
	
	
	<data id="sensorData" src="http://localhost:8080/sensordata.json"/>
        <data id="detectionparam" src="http://localhost:8080/sample.json" />
        <data id="dp1"/>
        <data id="dp2" />
		 <data id="dp3" expr="true"/>
		 <data id="dp4" />

    </datamodel>

 

    <state id="No_Fault">

 

        
        <onentry>

 
 

                    
             <assign location="dp1" expr="sensorData.HK_200300 &lt; 25" />
            <assign location="dp2" expr="true" />
			<assign location="dp4" expr="true"/>

            <log label="DP1 Value On Entry" expr="dp1" />
			            <log label="DP3 Value On Entry" expr="dp3" />

			<send  event="ec1"/>
			

        </onentry>

        <transition event="'ec1'" cond="dp1 or dp2 or dp3" target="Fault">
		
            <log label="DP1" expr="dp1" />
            <log label="DP2" expr="dp2" />
            <log label="DP3" expr="dp3" />
        </transition>

        <onexit>
            <log label="DP1 Value On Exit" expr="dp1" />
        </onexit>

 

    </state>

 

    <state id="Fault" initial="Root_Cause">

 

        <datamodel>
            <data id="rc1" />
            <data id="rc2"/>
            <data id="rc3"/>
			<data id="fm1"/>
        </datamodel>

 

        <parallel id="Root_Cause">

 <onentry>
 <assign location="rc1" expr="dp1 and dp2"/>

  <assign location="rc2" expr="dp2 and dp3"/>

 <assign location="rc3" expr="dp3"/>

 
 </onentry>

            <state id="RC1" initial="RC1_Initial">

 

                <state id="RC1_Initial">
                    <transition cond="rc1" target="rc1" />
                </state>

 

                <final id="rc1">
                <onentry>
                <log label="Fault RC1" expr="rc1" />
                </onentry>
                </final>

 

            </state>

 

 

            <state id="RC2" initial="RC2_Initial">

 

                <state id="RC2_Initial">
                    <transition cond="rc2" target="rc2" />
                </state>

 

                <final id="rc2">
                <onentry>
                <log label="Fault RC2" expr="rc2" />
                </onentry>
                </final>
                
            </state>

 

            <state id="RC3" initial="RC3_Initial">

 

                <state id="RC3_Initial">
                    <transition cond="rc3" target="rc3" />
                </state>

 

                <final id="rc3">
                <onentry>
                <log label="Fault RC3" expr="rc3" />
                </onentry>
                </final>

 

            </state>

 

            <transition target="Fault_Mode" />

 <onexit>
 <assign location="fm1" expr="rc1 and rc2"/>
 </onexit>

        </parallel>

 

 

        <parallel id="Fault_Mode">
            <state id="FM1" initial="FM1_Check">
                <state id="FM1_Check">
                <transition cond="fm1" target="fm1"/>
                </state>
                
                <final id="fm1">
                <onentry>
                <log label="Fault Mode FM1" expr="fm1" />
                </onentry>
                </final>
            </state>

 

            <state id="FM2">
                <onentry>
                    <log label="FM2" />
                </onentry>
            </state>
        </parallel>

 

    </state>
</scxml>

**Comment by alexzhornyak** [comment]
I do not see any **\<send\>** elements with attribute **'delay'** or **'delayexpr'** in your SCXML source code! If you mean another subject then explain detaily, please.

![image](https://user-images.githubusercontent.com/18611095/103280383-f0ffed80-49d8-11eb-8e6c-d4fc7c936504.png)

I offer you to use [Visual SCXML Editor](https://github.com/alexzhornyak/ScxmlEditor-Tutorial) that supports local debugging and test locally the behaviour of your statemachine.
You may open [Issue_5.zip](https://github.com/alexzhornyak/SCXML-tutorial/files/5750532/Issue_5.zip) with the editor.

### Recommended SCXML Code Standard
1) It is not recommended to use same names for states and data fields
![image](https://user-images.githubusercontent.com/18611095/103280578-6e2b6280-49d9-11eb-8b88-459fca8e1dca.png)

2) It is not recommended to use names for states that differ only by case
![image](https://user-images.githubusercontent.com/18611095/103281233-fd854580-49da-11eb-98af-02f186b721ea.png)

**Comment by seenivasanmca** [comment]
<scxml datamodel="jexl" location="Scxml" version="1.0"
    xmlns="http://www.w3.org/2005/07/scxml">

 


    <datamodel>
	
	
	<data id="sensorData" src="http://localhost:8080/sensordata.json"/>
        <data id="detectionparam" src="http://localhost:8080/sample.json" />
        <data id="dp1"/>
        <data id="dp2" />
		 <data id="dp3" expr="true"/>
		 <data id="dp4" />


    </datamodel>

 

    <state id="No_Fault">

 

        
        <onentry>

 
 

                    
             <assign location="dp1" expr="sensorData.HK_200300 &lt; 25" />
            <assign location="dp2" expr="true" />
			<assign location="dp4" expr="true"/>



            <log label="DP1 Value On Entry" expr="dp1" />
			            <log label="DP3 Value On Entry" expr="dp3" />

			<send  event="ec1" delay="5s"/>
			

        </onentry>


        <transition event="'ec1'" cond="dp1 or dp2 or dp3" target="Fault">
		
            <log label="DP1" expr="dp1" />
            <log label="DP2" expr="dp2" />
            <log label="DP3" expr="dp3" />
        </transition>

        <onexit>
            <log label="DP1 Value On Exit" expr="dp1" />
        </onexit>

 


    </state>

 

    <state id="Fault" initial="Root_Cause">

 

        <datamodel>
            <data id="rc1" />
            <data id="rc2"/>
            <data id="rc3"/>
			<data id="fm1"/>
        </datamodel>

 

        <parallel id="Root_Cause">

 <onentry>
 <assign location="rc1" expr="dp1 and dp2"/>

  <assign location="rc2" expr="dp2 and dp3"/>

 <assign location="rc3" expr="dp3"/>

 
 </onentry>

            <state id="RC1" initial="RC1_Initial">

 

                <state id="RC1_Initial">
                    <transition cond="rc1" target="rc1" />
                </state>

 

                <final id="rc1">
                <onentry>
                <log label="Fault RC1" expr="rc1" />
                </onentry>
                </final>

 


            </state>

 

 

            <state id="RC2" initial="RC2_Initial">

 

                <state id="RC2_Initial">
                    <transition cond="rc2" target="rc2" />
                </state>

 

                <final id="rc2">
                <onentry>
                <log label="Fault RC2" expr="rc2" />
                </onentry>
                </final>
                
            </state>

 

            <state id="RC3" initial="RC3_Initial">

 

                <state id="RC3_Initial">
                    <transition cond="rc3" target="rc3" />
                </state>

 

                <final id="rc3">
                <onentry>
                <log label="Fault RC3" expr="rc3" />
                </onentry>
                </final>

 

            </state>

 

            <transition target="Fault_Mode" />

 <onexit>
 <assign location="fm1" expr="rc1 and rc2"/>
 </onexit>

        </parallel>

 

 

        <parallel id="Fault_Mode">
            <state id="FM1" initial="FM1_Check">
                <state id="FM1_Check">
                <transition cond="fm1" target="fm1"/>
                </state>
                
                <final id="fm1">
                <onentry>
                <log label="Fault Mode FM1" expr="fm1" />
                </onentry>
                </final>
            </state>

 

            <state id="FM2">
                <onentry>
                    <log label="FM2" />
                </onentry>
            </state>
        </parallel>

 

    </state>
</scxml>

**Comment by seenivasanmca** [comment]
update scxml with delay attribute


 



<scxml datamodel="jexl" location="Scxml" version="1.0"
    xmlns="http://www.w3.org/2005/07/scxml">

 


    <datamodel>
	
	
	<data id="sensorData" src="http://localhost:8080/sensordata.json"/>
        <data id="detectionparam" src="http://localhost:8080/sample.json" />
        <data id="dp1"/>
        <data id="dp2" />
		 <data id="dp3" expr="true"/>
		 <data id="dp4" />


    </datamodel>

 

    <state id="No_Fault">

 

        
        <onentry>

 
 

                    
             <assign location="dp1" expr="sensorData.HK_200300 &lt; 25" />
            <assign location="dp2" expr="true" />
			<assign location="dp4" expr="true"/>



            <log label="DP1 Value On Entry" expr="dp1" />
			            <log label="DP3 Value On Entry" expr="dp3" />

			<send  event="ec1" delay="5s"/>
			

        </onentry>


        <transition event="'ec1'" cond="dp1 or dp2 or dp3" target="Fault">
		
            <log label="DP1" expr="dp1" />
            <log label="DP2" expr="dp2" />
            <log label="DP3" expr="dp3" />
        </transition>

        <onexit>
            <log label="DP1 Value On Exit" expr="dp1" />
        </onexit>

 


    </state>

 

    <state id="Fault" initial="Root_Cause">

 

        <datamodel>
            <data id="rc1" />
            <data id="rc2"/>
            <data id="rc3"/>
			<data id="fm1"/>
        </datamodel>

 

        <parallel id="Root_Cause">

 <onentry>
 <assign location="rc1" expr="dp1 and dp2"/>

  <assign location="rc2" expr="dp2 and dp3"/>

 <assign location="rc3" expr="dp3"/>

 
 </onentry>

            <state id="RC1" initial="RC1_Initial">

 

                <state id="RC1_Initial">
                    <transition cond="rc1" target="rc1" />
                </state>

 

                <final id="rc1">
                <onentry>
                <log label="Fault RC1" expr="rc1" />
                </onentry>
                </final>

 


            </state>

 

 

            <state id="RC2" initial="RC2_Initial">

 

                <state id="RC2_Initial">
                    <transition cond="rc2" target="rc2" />
                </state>

 

                <final id="rc2">
                <onentry>
                <log label="Fault RC2" expr="rc2" />
                </onentry>
                </final>
                
            </state>

 

            <state id="RC3" initial="RC3_Initial">

 

                <state id="RC3_Initial">
                    <transition cond="rc3" target="rc3" />
                </state>

 

                <final id="rc3">
                <onentry>
                <log label="Fault RC3" expr="rc3" />
                </onentry>
                </final>

 

            </state>

 

            <transition target="Fault_Mode" />

 <onexit>
 <assign location="fm1" expr="rc1 and rc2"/>
 </onexit>

        </parallel>

 

 

        <parallel id="Fault_Mode">
            <state id="FM1" initial="FM1_Check">
                <state id="FM1_Check">
                <transition cond="fm1" target="fm1"/>
                </state>
                
                <final id="fm1">
                <onentry>
                <log label="Fault Mode FM1" expr="fm1" />
                </onentry>
                </final>
            </state>

 

            <state id="FM2">
                <onentry>
                    <log label="FM2" />
                </onentry>
            </state>
        </parallel>

 

    </state>
</scxml>

**Comment by seenivasanmca** [comment]
updated scxml with delay attributes
 



<scxml datamodel="jexl" location="Scxml" version="1.0"
    xmlns="http://www.w3.org/2005/07/scxml">

 


    <datamodel>
	
	
	<data id="sensorData" src="http://localhost:8080/sensordata.json"/>
        <data id="detectionparam" src="http://localhost:8080/sample.json" />
        <data id="dp1"/>
        <data id="dp2" />
		 <data id="dp3" expr="true"/>
		 <data id="dp4" />


    </datamodel>

 

    <state id="No_Fault">

 

        
        <onentry>

 
 

                    
             <assign location="dp1" expr="sensorData.HK_200300 &lt; 25" />
            <assign location="dp2" expr="true" />
			<assign location="dp4" expr="true"/>



            <log label="DP1 Value On Entry" expr="dp1" />
			            <log label="DP3 Value On Entry" expr="dp3" />

			<send  event="ec1" delay="5s"/>
			

        </onentry>


        <transition event="'ec1'" cond="dp1 or dp2 or dp3" target="Fault">
		
            <log label="DP1" expr="dp1" />
            <log label="DP2" expr="dp2" />
            <log label="DP3" expr="dp3" />
        </transition>

        <onexit>
            <log label="DP1 Value On Exit" expr="dp1" />
        </onexit>

 


    </state>

 

    <state id="Fault" initial="Root_Cause">

 

        <datamodel>
            <data id="rc1" />
            <data id="rc2"/>
            <data id="rc3"/>
			<data id="fm1"/>
        </datamodel>

 

        <parallel id="Root_Cause">

 <onentry>
 <assign location="rc1" expr="dp1 and dp2"/>

  <assign location="rc2" expr="dp2 and dp3"/>

 <assign location="rc3" expr="dp3"/>

 
 </onentry>

            <state id="RC1" initial="RC1_Initial">

 

                <state id="RC1_Initial">
                    <transition cond="rc1" target="rc1" />
                </state>

 

                <final id="rc1">
                <onentry>
                <log label="Fault RC1" expr="rc1" />
                </onentry>
                </final>

 


            </state>

 

 

            <state id="RC2" initial="RC2_Initial">

 

                <state id="RC2_Initial">
                    <transition cond="rc2" target="rc2" />
                </state>

 

                <final id="rc2">
                <onentry>
                <log label="Fault RC2" expr="rc2" />
                </onentry>
                </final>
                
            </state>

 

            <state id="RC3" initial="RC3_Initial">

 

                <state id="RC3_Initial">
                    <transition cond="rc3" target="rc3" />
                </state>

 

                <final id="rc3">
                <onentry>
                <log label="Fault RC3" expr="rc3" />
                </onentry>
                </final>

 

            </state>

 

            <transition target="Fault_Mode" />

 <onexit>
 <assign location="fm1" expr="rc1 and rc2"/>
 </onexit>

        </parallel>

 

 

        <parallel id="Fault_Mode">
            <state id="FM1" initial="FM1_Check">
                <state id="FM1_Check">
                <transition cond="fm1" target="fm1"/>
                </state>
                
                <final id="fm1">
                <onentry>
                <log label="Fault Mode FM1" expr="fm1" />
                </onentry>
                </final>
            </state>

 

            <state id="FM2">
                <onentry>
                    <log label="FM2" />
                </onentry>
            </state>
        </parallel>

 

    </state>
</scxml>

**Comment by alexzhornyak** [comment]
@seenivasanmca follow the next issue support conditions, please:
1) attach **full scxml file** as **zip** or **txt** with all necessary xml tags such as **\<scxml\>** element
2) try **not to send** state chart with **syntax or non-standard errors** such as `<transition event="'ec1'"` (must be `<transition event="ec1"`) where event **ec1** is enclosed with **' character** and there is a lot of time wasted to find such error
3) check in any [visual SCXML editor](https://github.com/alexzhornyak/ScxmlEditor-Tutorial) the behaviour of your statemachine first. Is the behaviour as expected or not!? Because I do not know what is expected from your source code, sorry
4) read [the SCXML standard](https://github.com/alexzhornyak/SCXML-tutorial/blob/master/Doc/send.md) concerning the element you need. For example, attribute **'delayexpr'** in \<send\> element must return something like `return 25*45 + 'ms'`

Your present code is working something like that now
![Issue_5](https://user-images.githubusercontent.com/18611095/103286764-a044c080-49e9-11eb-93f8-ae1ef9304ce4.gif)
